runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker.cpp

```diff
@@ -0,0 +1,162 @@
+OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_initCifNativeThunkData(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	FFITypeHelpers FFIHelpers = FFITypeHelpers(currentThread);
+	ffi_cif *cif = NULL;
+	ffi_type *returnType = NULL;
+	ffi_type **argTypes = NULL;
+	J9CifArgumentTypes *cifArgTypesNode = NULL;
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	bool newArgTypes = (bool)(*(U_32*)currentThread->sp);
+	j9object_t retLayoutStrObject = J9_JNI_UNWRAP_REFERENCE(currentThread->sp + 1);
+	j9object_t argLayoutStrsObject = J9_JNI_UNWRAP_REFERENCE(currentThread->sp + 2);
+	j9object_t nativeInvoker = J9_JNI_UNWRAP_REFERENCE(currentThread->sp + 3);
+	U_32 argTypesCount = J9INDEXABLEOBJECT_SIZE(currentThread, argLayoutStrsObject);
+	returnType = (ffi_type *)FFIHelpers.getPrimitiveFFIType(retLayoutStrObject);
+
+	if (!newArgTypes) {
+		argTypes = (ffi_type **)(UDATA)J9VMJDKINTERNALFOREIGNABIPROGRAMMABLEINVOKER_ARGTYPESADDR(currentThread, nativeInvoker);
+	} else {
+		argTypes = (ffi_type **)j9mem_allocate_memory(sizeof(ffi_type *) * argTypesCount, OMRMEM_CATEGORY_VM);
+		if (NULL == argTypes) {
+			rc = GOTO_THROW_CURRENT_EXCEPTION;
+			setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto done;
+		}
+
+		memset(argTypes, 0, sizeof(ffi_type *) * argTypesCount);
+		for (U_8 i = 0; i < argTypesCount; i++) {
+			argTypes[i] = (ffi_type *)FFIHelpers.getPrimitiveFFIType(J9JAVAARRAYOFOBJECT_LOAD(currentThread, argLayoutStrsObject, i));
+		}
+	}
+
+	if (NULL == vm->cifNativeCalloutDataCache) {
+		vm->cifNativeCalloutDataCache = pool_new(sizeof(ffi_cif), 0, 0, 0, J9_GET_CALLSITE(), OMRMEM_CATEGORY_VM, POOL_FOR_PORT(PORTLIB));
+		if (NULL == vm->cifNativeCalloutDataCache) {
+			rc = GOTO_THROW_CURRENT_EXCEPTION;
+			setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto freeAllMemoryThenExit;
+		}
+	}
+
+	omrthread_monitor_enter(vm->cifNativeCalloutDataCacheMutex);
+	cif = (ffi_cif *)pool_newElement(vm->cifNativeCalloutDataCache);
+	omrthread_monitor_exit(vm->cifNativeCalloutDataCacheMutex);
+	if (NULL == cif) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto freeAllMemoryThenExit;
+	}
+
+	if (FFI_OK != ffi_prep_cif(cif, FFI_DEFAULT_ABI, argTypesCount, returnType, &(argTypes[0]))) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, NULL);
+		goto freeAllMemoryThenExit;
+	}
+
+	if (newArgTypes) {
+		cifArgTypesNode = (J9CifArgumentTypes *)j9mem_allocate_memory(sizeof(J9CifArgumentTypes), OMRMEM_CATEGORY_VM);
+		if (NULL == cifArgTypesNode) {
+			rc = GOTO_THROW_CURRENT_EXCEPTION;
+			setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto freeAllMemoryThenExit;
+		}
+
+		cifArgTypesNode->argumentTypes = (void **)argTypes;
+		omrthread_monitor_enter(vm->cifArgumentTypesMutex);
+		J9_LINKED_LIST_ADD_LAST(vm->cifArgumentTypesListHead, cifArgTypesNode);
```

Comment:
why is `cifArgumentTypesListHead` a linked list? it doesn't seem like we ever iterate it. Can it be a pool instead? Something with less memory overhead (e.g J9ThreadMonitorPool)

Reply:
Yes we never use the inserted node in the list anywhere in code as the only reason for this is to release all the memory in case of memory leak when VM exits unexpectedly. If pool works in the same way with less overhead, we should change to that.

Reply:
Replaced with `J9Pool *cifArgumentTypesCache` as suggested.